saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/serviceprovider/authentication/OpenSamlAuthenticationProvider.java

```diff
@@ -0,0 +1,332 @@
+		List<StatementValidator> statements = Collections.emptyList();
+
+		Set<Credential> credentials = new HashSet<>();
+		for (X509Certificate key : getVerificationKeys(provider)) {
+			final Credential cred = getVerificationCredential(key);
+			credentials.add(cred);
+		}
+		CredentialResolver credentialsResolver = new CollectionCredentialResolver(credentials);
+		SignatureTrustEngine signatureTrustEngine = new ExplicitKeySignatureTrustEngine(
+				credentialsResolver,
+				DefaultSecurityConfigurationBootstrap.buildBasicInlineKeyInfoCredentialResolver()
+		);
+		SignaturePrevalidator signaturePrevalidator = new SAMLSignatureProfileValidator();
+		return new SAML20AssertionValidator(
+				conditions,
+				subjects,
+				statements,
+				signatureTrustEngine,
+				signaturePrevalidator
+		);
+	}
+
+	private Credential getVerificationCredential(X509Certificate certificate) {
+		return CredentialSupport.getSimpleCredential(certificate, null);
+	}
+
+	private Decrypter getDecrypter(Saml2X509Credential key) {
+		Credential credential = CredentialSupport.getSimpleCredential(key.getCertificate(), key.getPrivateKey());
+		KeyInfoCredentialResolver resolver = new StaticKeyInfoCredentialResolver(credential);
+		Decrypter decrypter = new Decrypter(null, resolver, saml.getEncryptedKeyResolver());
+		decrypter.setRootInNewDocument(true);
+		return decrypter;
+	}
+
+	private Assertion decrypt(Saml2AuthenticationToken idp, EncryptedAssertion assertion) {
+		Saml2Exception last = null;
+		for (Saml2X509Credential key : getDecryptionCredentials(idp)) {
+			final Decrypter decrypter = getDecrypter(key);
+			try {
+				return decrypter.decrypt(assertion);
+			}
+			catch (DecryptionException e) {
+				throw new Saml2Exception(e);
+			}
+		}
+		throw last;
+	}
+
+	private NameID decrypt(Saml2AuthenticationToken token, EncryptedID assertion) {
+		Saml2Exception last = null;
+		for (Saml2X509Credential key : getDecryptionCredentials(token)) {
+			final Decrypter decrypter = getDecrypter(key);
+			try {
+				return (NameID) decrypter.decrypt(assertion);
+			}
+			catch (DecryptionException e) {
+				last = new Saml2Exception(e);
+			}
+		}
+		throw last;
+	}
+
+	private List<Saml2X509Credential> getDecryptionCredentials(Saml2AuthenticationToken token) {
+		return token.getX509Credentials()
+				.stream()
+				.filter(Saml2X509Credential::isDecryptionCredential)
+				.collect(Collectors.toList());
+	}
+
+	private List<X509Certificate> getVerificationKeys(Saml2AuthenticationToken token) {
+		return token.getX509Credentials()
+				.stream()
+				.filter(Saml2X509Credential::isSignatureVerficationCredential)
+				.map(c -> c.getCertificate())
+				.collect(Collectors.toList());
+	}
+
+	public void setAuthoritiesMapper(GrantedAuthoritiesMapper authoritiesMapper) {
+		notNull(authoritiesMapper, "authoritiesMapper must not be null");
+		this.authoritiesMapper = authoritiesMapper;
+	}
+
+	public void setResponseTimeValidationSkewMillis(int responseTimeValidationSkewMillis) {
```

Comment:
I think this would be better as a `Duration`, as that aligns with `JwtTimestampValidator`, `OAuth2AuthorizedClientBuilder`, and a few others.

Additionally, it appears the code translates it into a `Duration` anyway.

Reply:
Agreed. Resolved